📖 document providers conventions for clusterctl (#1893)
kubernetes-sigs/cluster-api: docs/book/src/providers/clusterctl.md

# `clusterctl`

## clusterctl v1alpha3 (clusterctl redesign)

`clusterctl` is a CLI tool for handling the lifecycle of a cluster-API management cluster.

> The v1alpha3 release is designed for providing a simple a day 1 experience; `clusterctl` is bundled with Cluster API and can be reused across providers

Reviewer: ```suggestion
The v1alpha3 release is designed for providing a simple day 1 experience; `clusterctl` is bundled with Cluster API and can be reused across providers
```

Author: done

that are compliant with the following rules.

### Components YAML

The provider is required to generate a single YAML file with all the components required for installing the provider
itself (CRD, Controller, RBAC etc.). 

Infrastructure providers MUST release a file called `infrastructure-components.yaml`, while bootstrap provider MUST
release a file called ` bootstrap-components.yaml` (exception for CABPK, which is included in CAPI by default).

> The components YAML should contain one Namespace object only; this namespace will be used as a default target namespace

Reviewer: ```suggestion
The components YAML should contain one Namespace object, which will be used as the default namespace
```

Author: done

> for installing the provider components.

Reviewer: ```suggestion
when creating provider components.
```

Author: done

> > If the generated component YAML does not contain a Namespace object, `clusterctl` can determine the default namespace;

Reviewer: ```suggestion
> If the generated component YAML doesn't contain a Namespace object, users will need to provide one.
```

Author: done

> in this case the user will be always required to specify a target namespace for installing the provider.

> In case there is more than one Namespace object in the components YAML, `clusterctl` will generate an error and abort
> the provider installation.

The components YAML can contain environment variables matching the regexp `\${\s*([A-Z0-9_]+)\s*}`; it is highly
recommended to prefix the variable name with the provider name e.g. `{ $AWS_CREDENTIALS }`

> > The users should ensure those variables are set in advance before running `clusterctl`; in case some variable

Reviewer: ```suggestion
> Users are required to ensure that environment variables are set in advance before running `clusterctl`; if a variable
```

Author: done

> > are missing at run time, `clusterctl` will generate an error and abort the provider operation.

Reviewer: ```suggestion
> is missing, `clusterctl` will generate an error and abort the provider operation.
```

Author: done

> ### Cluster templates

Reviewer: ```suggestion
### Workload cluster templates
```

Author: done

Infrastructure provider could publish cluster templates to be used by `clusterctl config cluster`.

Cluster templates MUST be stored in the same folder of the component YAML and adhere to the the following naming convention:
1. The default cluster template should be named `config-{bootstrap}.yaml`. e.g `config-kubeadm.yaml`
2. Additional cluster template should be named `config-{flavor}-{bootstrap}.yaml`. e.g `config-production-kubeadm.yaml`

`{bootstrap}` is the name of the bootstrap provider used in the template; `{flavor}` is the name the user can pass to the
`clusterctl config cluster --flavor` flag to identify the specific template to use.

> ## Old versions 

Reviewer: ```suggestion
## Previous versions (unsupported) 
```

Author: done
